subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/install/internal/AdoptOpenJdkRemoteBinary.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.toolchain.install.internal;
+
+import com.google.common.io.Files;
+import net.rubygrapefruit.platform.SystemInfo;
+import org.apache.commons.io.FileUtils;
+import org.gradle.api.GradleException;
+import org.gradle.api.JavaVersion;
+import org.gradle.internal.os.OperatingSystem;
+import org.gradle.jvm.toolchain.JavaToolchainSpec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+public class AdoptOpenJdkRemoteBinary {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AdoptOpenJdkRemoteBinary.class);
+
+    private final SystemInfo systemInfo;
+    private final OperatingSystem operatingSystem;
+
+    @Inject
+    public AdoptOpenJdkRemoteBinary(SystemInfo systemInfo, OperatingSystem operatingSystem) {
+        this.systemInfo = systemInfo;
+        this.operatingSystem = operatingSystem;
+    }
+
+    // TODO: [bm] this might be downloading the same jdk multiple times in parallel builds
+    // TODO: [bm] Alternative is using RepositoryTransportFactory+ExternalResourceAccessor+ChecksumService
+    // Needs to be tied up with a FileStore impl as well
+    // Adds risk as AdoptOpenJDK is hosted on S3 which does not allow HEAD requests; they
+    // currently have a workaround to support HEAD requests for the "Gradle" User agent
+    // Not using accessor might help us to move all of this into jvmServices for TD reuse
+    public Optional<File> download(JavaToolchainSpec spec) {
+        if (!canProvidesMatchingJdk(spec)) {
+            return Optional.empty();
+        }
+        File tmpFile = new File(Files.createTempDir(), toFilename(spec));
+        int defaultTimeout = (int) TimeUnit.MINUTES.toMillis(5);
+        URL source = toDownloadUrl(spec);
+        LOGGER.info("Downloading {} to {}", source, tmpFile);
```

Comment:
:x: The download happens during configuration time. Which I think is ok, as it is cached after that. Although it would be "nicer" if we could do it later just before the JDK is used for the first time.

In any case, this should use `LOGGER.lifecycle` (need to use Logging.getLogger(...) for the Gradle specific logger). Otherwise I don't get any feedback what is going on. As a follow up, this should instead use progress logging in some form.
And maybe remove the "to /tmp/..." part from the log message as it only shows a tmp folder and not the final location.

Reply:
Hm why do you think this happens during configuration time? It depends when the first caller access the `JavaCompiler` on the `JavaCompile` task. Can have a look into this as well but it should only resolve when needed.

Reply:
It's resolved here:
https://github.com/gradle/gradle/blob/fdadc10d50a6cd17446b2cc333e9624534d6b835/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java#L187

Reply:
So one part to this is that the toolchain got resolved at task graph calculation time because it was used as an Input. I think there are some more things to tackle whether toolchains should count as input or which parts of them. I opened #13970 to track that

Reply:
Given that a follow up will add progress logging, we can keep the logging as it is here.